### PR TITLE
Enforce that the current realm is connected while running GATT procedures.

### LINKED ISSUE #114: Describe what happens in response to disconnection (a lost ATT Bearer)
https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothdevice-gattserver describes `BluetoothDevice.gattServer` becoming `null`, and [`BluetoothGATTRemoteServer.connected`](https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothgattremoteserver-connected) needs to become false.

I'm not certain that `BluetoothDevice.gattServer` should actually become null after the first connection, and if it does then maybe it should go back to the original object on reconnection?

### PR DESCRIPTION
**Pull request by jyasskin**
This change makes it so that a disconnect() call takes effect immediately and
prevents any subsequent changes to member variables from a returning GATT
procedure, even if the GATT procedure started before the disconnection. It also
forces all active requests to fail with a NetworkError so that no remnants of
the connection keep running after it's terminated.

This reverses the readValue() promise and event order from #175, in order to
avoid needing to perform a microtask checkpoint inside the algorithm.

Fixes #114 and #171.

Preview at https://rawgit.com/jyasskin/web-bluetooth-1/enforce-connectedness/index.html.

### CODE REVIEW

**Comment by jyasskin** [comment]
@yrliou @BruceSunMozilla this PR affects how you'll implement disconnection.

File: index.bs [hunk 1]
```diff
@@ -2196,6 +2244,65 @@ spec: webidl
     </ol>
 
     <p>
+      <span class="note">
+        Algorithms need to fail if
+        their {{BluetoothRemoteGATTServer}} was disconnected while they were running,
+        even if the UA stays connected the whole time
+        and the {{BluetoothRemoteGATTServer}} is subsequently re-connected before they finish.
+        We wrap the returned {{Promise}} to accomplish this.
+      </span>
```
Comment (g-ortuno) [inline index.bs]: Should we put the note in a separate paragraph?

File: index.bs [hunk 2]
```diff
@@ -2196,6 +2244,65 @@ spec: webidl
...
+      To create a <var>gattServer</var>-<dfn>connection-checking wrapper</dfn>
```
Comment (g-ortuno) [inline index.bs]: Do you mean `gatt`? We don't have `gattServer` anymore.

File: index.bs [hunk 3]
```diff
@@ -2196,6 +2244,65 @@ spec: webidl
...
+      around a {{Promise}} <var>promise</var>, the UA MUST:
```
Comment (g-ortuno) [inline index.bs]: I wonder if we really need this wrapper. At first it makes sense, this way we don't have to check `connected` in each function. But now, looking a the functions, it makes less sense. We basically check `connected` in all but one of them. I think explicitly checking `connected` makes more sense that the wrapper.

File: index.bs (hunk 2 above)
Comment (jyasskin) [inline index.bs]: We removed `BluetoothDevice.gattServer`, but `BluetoothDevice.gatt` still refers to a gatt server, so I like using the longer name for the variable.

File: index.bs (hunk 3 above)
Comment (jyasskin) [inline index.bs]: The wrapper manages adding and removing `promise` from the [`[[activeAlgorithms]]`](https://rawgit.com/jyasskin/web-bluetooth-1/enforce-connectedness/index.html#dom-bluetoothremotegattserver-activealgorithms-slot) slot, in addition to effectively checking `connected`. 
Repeating the "add to `[[activeAlgorithms]]`" and "if not in `[[activeAlgorithms]]` fail with `NetworkError`" text isn't so bad, but I wouldn't want to repeat the removal code in each algorithm.
However, I could just leave promises in `[[activeAlgorithms]]` in the spec, and rely on the implementation to optimize away the memory leak. What do you think?

Comment (g-ortuno) [inline index.bs]: Do we need [[activeAlgorithms]] if we manually check `connected` in each function?

Comment (jyasskin) [inline index.bs]: Yes; its purpose is to catch pages that disconnect and reconnect while an algorithm is in progress. I tried to say this in the note just before https://rawgit.com/jyasskin/web-bluetooth-1/enforce-connectedness/index.html#connection-checking-wrapper; do you see a way to make it clearer?

File: index.bs (hunk 1 above)
Comment (jyasskin) [inline index.bs]: Done.

**Comment by g-ortuno** [comment]
lgtm

**Comment by Emill** [comment]
As discussed in https://github.com/WebBluetoothCG/web-bluetooth/issues/274#issuecomment-268266639, I don't like the order "fire event _then_ resolve promise" since that makes it impossible to distinguish notifications from read responses.

**Pull request merged**